### DISCUSSION TITLE
Is there a way to unit test composables?

### DISCUSSION THREAD

**Discussion by harm-matthias-harms** [nuxt/nuxt#16206] [Questions]
Is there currently a way to unit test `useState` composables with vitest?

```ts
// useFoo.ts
export const useFoo = () => {
  return useState("foo", () => "bar");
};
```

#### Naive idea 1:
```ts
// useFoo.test.ts
import { describe, expect, it } from "vitest";
import { useFoo } from "./useFoo"

describe("useFoo", () => {
  it("knows useFoo", () => {
    expect(useFoo().value).toEqual("bar");
  });
}

// => useState unknown
```
if I import useState and make an alias for "#app" it errors with `Error: nuxt instance unavailable`

#### Naive idea 2:
```ts
import { fileURLToPath } from "url";
import { describe, expect, it } from "vitest";
import { setup, $fetch } from "@nuxt/test-utils";

describe("example", async () => {
  await setup({
    rootDir: fileURLToPath(new URL("..", import.meta.url)),
    server: true,
  });

  it("knows useFoo", () => {
    expect(useFoo().value).toEqual("bar");
  });
});

// => useFoo is not defined
// importing useFoo and useState will result in the same error as idea 1
```

As mentioned above those are stupid ideas, each having a reasonable error. Still, I'm searching for a way to test the business logic without testing the html over the $fetch method. And without making stores, such as pinia, necessary, just for the unit testing.

**Comment by thegouge** [comment]
I'm also in this camp in that I have SEVERAL composables that I've built for a project that fetch external data via useFetch, using the config from useRuntimeConfig and would love to be able to test my composables.

so I wouldn't call the request in here stupid, there should be a way to test code that depends on Nuxt's built-in composables!

**Comment by huang-julien** [comment]
using ``vitest.mock` doesn't work ? I usually mock #imports instead of #app

**Comment by valentineus** [comment]
This question is really important. We've made an internal API and we call methods with composables. And we don't have the tools to test these functions.

We can test the server API endpoints. But we can't test the method call interface of composables. And we've already found bugs that could have been avoided by testing.

**Comment by thegouge** [comment]
The workaround for this that I've cobbled together is that I'm using vitest, and resolving "#imports" to a mock file that I've built

for clarification: 
```
// vitest.config.ts

export default defineConfig({
  resolve: {
    alias: {
      '~': __dirname,
      '#imports': resolve(__dirname, './imports-mock.ts'),
    },
  },
...
})
```

and then I built out mocked versions of Nuxt's internal composables like so:

```
// imports-mock.ts

import fetch from 'node-fetch'
import { ref } from 'vue'

export const mockServerPath = 'http://localhost:8000'
let path = '/'

export function useRoute() {
  return { path }
}

export function useRouter() {
  return {
    push: (newPath) => {
      path = newPath
    },
  }
}

export function useRuntimeConfig() {
  return {
    public: {
      // mocked versions of all my environment variables
    },
  }
}

export async function useFetch(url, options) {
  try {
    const response = await fetch(
      `${url.includes('http') ? '' : `${mockServerPath}`}${url}`,
      {
        ...options,
      }
    )

    let data = ref(await response.json())

    if (options?.transform) {
      data = ref(options.transform(data.value))
    }

    return { data, error: false, pending: false, refresh: () => ({}) }
  } catch (error) {
    return {
      data: null,
      error,
      pending: false,
      refresh: () => ({}),
    }
  }
}

export function useState(name, init = () => null) {
  return init()
}

export function useCookie(name, init = () => null) {
  return init()
}
```

my `imports-mock` is kind of a work in progress that I'm adding to as I need while I build out vitest unit tests for all my composables, but this setup has proved pretty effective for me so far (in conjunction with vitest.mock). I hope this helps!

**Reply by valentineus** [reply]
Your version looks like an interesting solution.
I will definitely try to implement it!

**Comment by edumudu** [comment]
Use [nuxt-vitest](https://github.com/danielroe/nuxt-vitest), I am using it and its working so far
